superset-frontend/src/explore/components/controls/DateFilterControl/DateFilterControl.tsx

```diff
@@ -0,0 +1,872 @@
+    }
+    if (!value.includes(SEPARATOR) && value.startsWith('Last')) {
+      since = value;
+    }
+    if (!value.includes(SEPARATOR) && value.startsWith('Next')) {
+      until = value;
+    }
+    return `${since}${SEPARATOR}${until}`;
+  }
+
+  function onAdvancedRangeChange(control: 'since' | 'until', value: string) {
+    setValidTimeRange(false);
+    setEvalTimeRange(t('Need to verify the time range.'));
+    const [since, until] = advancedRange.split(SEPARATOR);
+    if (control === 'since') {
+      setAdvancedRange(`${value}${SEPARATOR}${until}`);
+    } else {
+      setAdvancedRange(`${since}${SEPARATOR}${value}`);
+    }
+  }
+
+  function onCustomRangeChange(
+    control: CustomRangeKey,
+    value: string | number,
+  ) {
+    setCustomRange({
+      ...customRange,
+      [control]: value,
+    });
+  }
+
+  function onCustomRangeChangeAnchorMode(option: any) {
+    const radioValue = option.target.value;
+    if (radioValue === 'now') {
+      setCustomRange({
+        ...customRange,
+        anchorValue: 'now',
+        anchorMode: radioValue,
+      });
+    } else {
+      setCustomRange({
+        ...customRange,
+        anchorValue: DEFAULT_UNTIL,
+        anchorMode: radioValue,
+      });
+    }
+  }
+
+  function showValidateBtn(): boolean {
+    return timeRangeFrame === 'Advanced';
+  }
+
+  function resetState(value: string) {
+    setTimeRangeFrame(guessTimeRangeFrame(value));
+    setCommonRange(getDefaultOrCommonRange(value));
+    setCalendarRange(getDefaultOrCalendarRange(value));
+    setCustomRange(customTimeRangeDecode(value).customRange);
+    setAdvancedRange(getAdvancedRange(value));
+    setShow(false);
+  }
+
+  function onSave() {
+    const currentValue = getCurrentValue();
+    onChange(currentValue);
+    resetState(currentValue);
+  }
+
+  function onHide() {
+    resetState(value);
+  }
+
+  function onValidate() {
+    const value = getCurrentValue();
+    fetchTimeRange(value, endpoints).then(({ value, error }) => {
+      if (error) {
+        setEvalTimeRange(error || '');
+        setValidTimeRange(false);
+      } else {
+        setEvalTimeRange(value || '');
+        setValidTimeRange(true);
+      }
+    });
+  }
+
+  function renderCommon() {
+    const commonRangeValue =
+      COMMON_RANGE_OPTIONS.find(_ => _.value === commonRange)?.value ||
+      'Last week';
+    return (
+      <>
+        <div className="section-title">
+          {t('Configure Time Range: Last...')}
+        </div>
+        <Radio.Group
+          value={commonRangeValue}
+          onChange={(e: any) => setCommonRange(e.target.value)}
+        >
+          {COMMON_RANGE_OPTIONS.map(_ => (
+            <Radio key={_.value} value={_.value} className="vertical-radio">
+              {_.label}
+            </Radio>
+          ))}
+        </Radio.Group>
+      </>
+    );
+  }
+
+  function renderCalendar() {
+    const currentValue =
+      CALENDAR_RANGE_OPTIONS.find(_ => _.value === calendarRange)?.value ||
```

Comment:
```suggestion
      CALENDAR_RANGE_OPTIONS.find(({ value }) => value === calendarRange)?.value ||
```

Reply:
thank you remind. I will use the destructuring syntax.